libs/custom-scanner/src/custom_a4_scanner.ts

```diff
@@ -0,0 +1,632 @@
+    } = {}
+  ): Promise<Result<T, ErrorCode>> {
+    let result!: Result<T, ErrorCode>;
+
+    for (let i = 0; i < max + 1; i += 1) {
+      debug('withRetries attempt %d/%d', i + 1, max + 1);
+
+      result = await fn();
+      if (result.isOk()) {
+        debug('withRetries success on attempt %d/%d', i + 1, max + 1);
+        break;
+      }
+
+      const error = result.err();
+      debug('withRetries error on attempt %d/%d: %o', i + 1, max + 1, error);
+      if (error !== ErrorCode.JobNotValid) {
+        break;
+      }
+    }
+
+    return result;
+  }
+
+  /**
+   * Moves a sheet of paper as directed.
+   */
+  async move(movement: FormMovement): Promise<Result<void, ErrorCode>> {
+    debug('moving %o', movement);
+    const result = await this.publicApiMutex.withLock(() =>
+      this.withRetries(async () => {
+        const createJobResult = await this.createJobInternal();
+        if (createJobResult.isErr()) {
+          return createJobResult;
+        }
+
+        return await this.channelMutex.withLock((channel) =>
+          formMove(channel, this.currentJobId, movement)
+        );
+      })
+    );
+    debug('move result: %o', result);
+    return result;
+  }
+
+  /**
+   * Scans a sheet of paper and returns the resulting images. Note that this
+   * method will always return a pair of `ImageFromScanner` objects, even if the
+   * scanner only scanned one side of the paper. The image buffer for the side
+   * that was not scanned will be empty.
+   */
+  async scan(
+    scanParameters: ScanParameters,
+    {
+      /* istanbul ignore next */
+      maxTimeoutNoMoveNoScan = 5_000,
+      /* istanbul ignore next */
+      maxRetries = 3,
+    }: { maxTimeoutNoMoveNoScan?: number; maxRetries?: number } = {}
+  ): Promise<Result<SheetOf<ImageFromScanner>, ErrorCode>> {
+    const timer = time(debug, 'scan');
+
+    timer.checkpoint('requesting lock');
+    const { unlock } = await this.publicApiMutex.asyncLock();
+    timer.checkpoint('got lock');
+
+    const { wantedScanSide: scanSide } = scanParameters;
+    const scannerImageA: ImageFromScanner = {
+      imageBuffer: Buffer.alloc(0),
+      imageWidth: 0,
+      imageHeight: 0,
+      imageDepth: scanParameters.imageColorDepth,
+      imageFormat: ImageFileFormat.Jpeg,
+      scanSide: ScanSide.A,
+      imageResolution: scanParameters.resolution,
+    };
+    const scannerImageB: ImageFromScanner = {
+      imageBuffer: Buffer.alloc(0),
+      imageWidth: 0,
+      imageHeight: 0,
+      imageDepth: scanParameters.imageColorDepth,
+      imageFormat: ImageFileFormat.Jpeg,
+      scanSide: ScanSide.B,
+      imageResolution: scanParameters.resolution,
+    };
+
+    timer.checkpoint('set scan parameters');
+    let setScanParametersResult = await this.setScanParametersInternal(
+      scanParameters
+    );
+
+    if (setScanParametersResult.isErr()) {
+      if (setScanParametersResult.err() === ErrorCode.JobNotValid) {
+        debug('job not valid, recreating job');
```

Comment:
Curious again why this happens and why we need a single retry. Maybe will have been answered in my previous comment about a similar case

Reply:
Again, because that's what Custom's example code did.